server/utils/firebase_algolia_server_code/functions/test/index.test.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * Description: Unit tests for index.js
+ *              Uses dependency injection to mock algolia save/delete functions
+ *              Uses sinon/chai to spy on calls and validate behavior
+ *              compile with `npm run test`
+ * Author: Robert Marcus
+ * Date: July 7, 2020
+ */
+
+// Import the testing environment configuration
+const { test, assert, sinon } = require('./testConfig');
+
+// Import the generic cloud functions.
+const { createRecordFromEntity, updateRecord, deleteRecord } = 
+    require('../index.js');
+
+// Import mock saveObject, mock deleteObject, and helpers.
+const { saveObject, deleteObject } = require('./algoliaMocks');
+const { getExpectedOutputID, getExpectedOutputSnap, getExpectedOutputChange } =
+    require('./expectedMockOutputHelpers');
+
+describe("test_createRecordFromEntity", () => {
+  it('should call saveObject exactly once', () => {
+    const saveSpy = sinon.spy(saveObject);
+    const snap = test.firestore.exampleDocumentSnapshot();
+
+    const createRecordWrapper = test.wrap(createRecordFromEntity(saveSpy));
+    const addedSnap = createRecordWrapper(snap);
+
+    assert.isTrue(saveSpy.calledOnce);
+  });
+
+  it('should call saveObject with the correct parameters', () => {
+    const snap = test.firestore.exampleDocumentSnapshot();
+    const saveSpy = sinon.spy(saveObject);
+
+    const createRecordWrapper = test.wrap(createRecordFromEntity(saveSpy));
+    const addedSnap = createRecordWrapper(snap);
+
+    assert.isTrue(saveSpy.calledWith(getExpectedOutputSnap(snap)));
+  });
+  
+  it('should return input data + "altered" field', () => {
+    const snap = test.firestore.exampleDocumentSnapshot();
+
+    const createRecordWrapper = test.wrap(createRecordFromEntity(saveObject));
+    const addedSnap = createRecordWrapper(snap);
+
+    assert.deepEqual(addedSnap, getExpectedOutputSnap(snap));
```

Comment:
As a general rule in writing tests, you want it to be very explicit what are the inputs and expected outputs.
That's not obvious at all in this test. What fields do `snap` and `addedSnap` contain exactly. What is the "altered field" that's added. I can't tell.

Reply:
`snap` is I (believe) a randomly generated `firestore snapshot`. I wouldn't want to pick it apart and write that example out manually as it's large and I'd probably get it wrong. To clarify what's new between the two I will, as you suggested, ignore DRY and make it as explicit as conveniently possible.

Reply:
DRY is less important in tests.  Tests are entirely independent of each other and each should be a completely readable unit in and of itself.  I want to be able to look at a test and tell exactly what's going on, what you are doing and what you are expecting the outcome to be, with minimal context.  This will be true for future engineers who write code that breaks your tests and then have to go understand why the test is breaking.

Imagine you're writing some code and discover that it breaks some tests in some part of the codebase you've never interacted with before.  If you can just go look at those tests and the line that's failing and understand clearly what the test is doing and what it's expecting to happen, it's straightforward to discover why your change is breaking that assumption.  If you have to go jump from file to file and understand the entire codebase to get to that point, you'll spend all day doing it and the rest of the year complaining about that test suite to anyone who will listen.